### PR TITLE
Store

### LINKED ISSUE #14: Implement storing of wallet
Implement `storage_handler` - methods `store` and `load`
 * `store` takes members added in #8 as arguments
 * Individual `std::string`s will be serialised to single `0`-separated `char *` before encryption
 * Add ocall function `store` 
     - Will take encrypted array of  bytes
     - File will be stored in same location as `wuss` executable
 * Add ocall function `load` - read encrypted file
 * Implement `load` method 
     * will get encrypted bytes via ocall `load` 
     * performs decryption and deserialization
     * Will return data as defined in #8

### PR DESCRIPTION
**Pull request by tajriis**
Closes #14

### CODE REVIEW

File: trusted/wallet.cpp [hunk 1]
```diff
@@ -35,11 +31,18 @@ wallet& wallet::get_instance()
 
 bool wallet::create_wallet(const password_t& mp_)
 {
+    int ret;
+    sgx_status_t status = wallet_exists(&ret);
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Is it ok that the return value is not used?
If yes, do we have to store it?

File: trusted/wallet.cpp [hunk 2]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+        output_ = std::copy(item.username.begin(), item.username.end(), output_);
+        output_++;
+        output_ = std::copy(item.password.begin(), item.password.end(), output_);
+        output_++;
+    }
+
+    sgx_status_t sealing_status = seal_wallet(data.get(), wallet_size, (sgx_sealed_data_t*)sealed_data.get(), sealed_size);
+
+    if (sealing_status != SGX_SUCCESS){
+        on_error("[update_stored_wallet] Wallet not sealed");
```
Comment (mdugacek) [inline trusted/wallet.cpp]: I think that `[update_stored_wallet] Failed to seale wallet`  would be better error message.

File: trusted/wallet.cpp [hunk 3]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+    }
+
+    int ret;
+    sgx_status_t stored_status = store_wallet(&ret, sealed_data.get(), sealed_size);
+    if (ret != 0 || stored_status != SGX_SUCCESS){
+        on_error("[update_stored_wallet] Wallet not stored to file");
```
Comment (mdugacek) [inline trusted/wallet.cpp]: I think that `[update_stored_wallet] Failed to store wallet` would be better error message.

File: trusted/wallet.cpp [hunk 4]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+    
+    size_t wallet_size = sealed_size - sizeof(sgx_sealed_data_t);
+    uint32_t loaded_size = wallet_size;
+    std::unique_ptr<uint8_t[]> sealed_data(new uint8_t[sealed_size]);
+    std::unique_ptr<uint8_t[]> unsealed_data(new uint8_t[wallet_size]);
+ 
+    int ret;
+    bool stored_status = load_wallet(&ret, sealed_data.get(), sealed_size);
+    if (ret != 0 || stored_status != SGX_SUCCESS){
+        on_error("[load_stored_wallet] Wallet not loaded from file");
```
Comment (mdugacek) [inline trusted/wallet.cpp]: I think that `[load_stored_wallet] Failed to load wallet from file` is better error message.

File: trusted/wallet.cpp [hunk 5]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+    }
+    sgx_status_t sealing_status = unseal_wallet((sgx_sealed_data_t*)sealed_data.get(), unsealed_data.get(), &loaded_size);
+    if (sealing_status != SGX_SUCCESS){
+        on_error("[load_stored_wallet] Wallet not unsealed");
```
Comment (mdugacek) [inline trusted/wallet.cpp]: I think that `[load_stored_wallet] Failed to open wallet` is better error message.

File: trusted/wallet.hpp [hunk 6]
```diff
@@ -47,11 +48,14 @@ class wallet
...
 public:
     void on_error(const std::string& message_) const;
+    void update_stored_wallet() const;
+    bool load_stored_wallet();
 
     // Private methods
 private:
-    void update_stored_wallet() const;
-
+    //void update_stored_wallet() const;
```
Comment (mdugacek) [inline trusted/wallet.hpp]: Delete comment.

File: trusted/wallet.hpp [hunk 7]
```diff
@@ -47,11 +48,14 @@ class wallet
...
+    //void load_stored_wallet() const;
```
Comment (mdugacek) [inline trusted/wallet.hpp]: Delete comment.

File: trusted/wallet.cpp [hunk 8]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+        return false;
+    }
+    
+    size_t wallet_size = sealed_size - sizeof(sgx_sealed_data_t);
+    uint32_t loaded_size = wallet_size;
+    std::unique_ptr<uint8_t[]> sealed_data(new uint8_t[sealed_size]);
+    std::unique_ptr<uint8_t[]> unsealed_data(new uint8_t[wallet_size]);
+ 
+    int ret;
+    bool stored_status = load_wallet(&ret, sealed_data.get(), sealed_size);
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Could be `const auto` and renamed to `load_success`.

File: trusted/wallet.cpp [hunk 9]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+bool wallet::load_stored_wallet()
+{
+    size_t sealed_size;
+    sgx_status_t status = get_file_size(&sealed_size);
+    if (sealed_size == 0)
+    {
+        return false;
+    }
+    
+    size_t wallet_size = sealed_size - sizeof(sgx_sealed_data_t);
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Add `const`.

File: trusted/wallet.cpp [hunk 10]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+    sgx_status_t stored_status = store_wallet(&ret, sealed_data.get(), sealed_size);
+    if (ret != 0 || stored_status != SGX_SUCCESS){
+        on_error("[update_stored_wallet] Wallet not stored to file");
+    }  
 }
 
+bool wallet::load_stored_wallet()
+{
+    size_t sealed_size;
+    sgx_status_t status = get_file_size(&sealed_size);
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Type could be `const auto`.

File: trusted/wallet.cpp [hunk 11]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
 
 void wallet::update_stored_wallet() const
 {
-    // TODO Save new stuff to file
+    size_t wallet_size = get_wallet_total_size() + _master_password.size() + 1;
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Could be `const auto` instead of `size_t`

File: trusted/wallet.cpp [hunk 12]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+    size_t sealed_size = wallet_size + sizeof(sgx_sealed_data_t);
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Could be `const auto` instead of `size_t`

File: trusted/wallet.hpp [hunk 13]
```diff
@@ -47,11 +48,14 @@ class wallet
     // OCALL Handlers
 public:
     void on_error(const std::string& message_) const;
+    void update_stored_wallet() const;
+    bool load_stored_wallet();
 
     // Private methods
 private:
```
Comment (mdugacek) [inline trusted/wallet.hpp]: This `private` section is empty so I think that you can delete it.

File: trusted/wallet.cpp [hunk 14]
```diff
@@ -215,7 +223,81 @@ void wallet::on_error(const std::string& message_) const
...
+
+    std::unique_ptr<uint8_t[]> sealed_data(new uint8_t[sealed_size]);
+
+    std::unique_ptr<uint8_t[]> data(new uint8_t[sealed_size]);
+
+    auto* output_ = data.get();
+    output_ = std::copy(_master_password.begin(), _master_password.end(), output_);
+    output_++;
+    for (const item_t& item : _items)
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Type could be `const auto`.

**Review by mdugacek — CHANGES_REQUESTED** [review]

File: misc/enclave.edl [hunk 15]
```diff
@@ -25,5 +25,13 @@ enclave {
 
     untrusted {
         void on_error([in, string]const char* message_);
+        size_t get_file_size();
+        int store_wallet([in, size=sealed_size]const uint8_t* sealed_data,
+                            size_t sealed_size);
+        int load_wallet([out, size=sealed_size]uint8_t* sealed_data,
+                        size_t sealed_size);
+
+        int wallet_exists(void);
```
Comment (JoyOfMadness) [inline misc/enclave.edl]: I don't think this is needed, we have `get_file_size` which returns 0 if file doesn't exist

File: trusted/storage_handler.cpp [hunk 16]
```diff
@@ -1,3 +1,13 @@
 #include "storage_handler.hpp"
+#include "sgx_trts.h"
```
Comment (JoyOfMadness) [inline trusted/storage_handler.cpp]: I don't think this header is needed

File: trusted/storage_handler.cpp [hunk 17]
```diff
@@ -1,3 +1,13 @@
...
+#include "sgx_tseal.h"
 
-// TODO
+sgx_status_t seal_wallet(const uint8_t* wallet_, size_t wallet_size, sgx_sealed_data_t* sealed_data, size_t sealed_size )
```
Comment (JoyOfMadness) [inline trusted/storage_handler.cpp]: For the sake of consistency, all function arguments should end with `_`, i.e. `wallet_size_`, `sealed_data_` ...

File: trusted/storage_handler.hpp [hunk 18]
```diff
@@ -1,3 +1,9 @@
 #pragma once
 
-// TODO
+#include "wallet.hpp"
+#include "sgx_trts.h"
```
Comment (JoyOfMadness) [inline trusted/storage_handler.hpp]: Again, probably not needed

File: trusted/wallet.cpp [hunk 19]
```diff
@@ -215,7 +223,80 @@ void wallet::on_error(const std::string& message_) const
...
 void wallet::update_stored_wallet() const
 {
-    // TODO Save new stuff to file
+    const auto wallet_size = get_wallet_total_size() + _master_password.size() + 1;
+    const auto sealed_size = wallet_size + sizeof(sgx_sealed_data_t);
+
+    std::unique_ptr<uint8_t[]> sealed_data(new uint8_t[sealed_size]);
+    std::unique_ptr<uint8_t[]> data(new uint8_t[sealed_size]);
+
+    auto* output_ = data.get();
```
Comment (JoyOfMadness) [inline trusted/wallet.cpp]: Ordinary variable shouldn't have trailing `_`

File: trusted/wallet.cpp [hunk 20]
```diff
@@ -215,7 +223,80 @@ void wallet::on_error(const std::string& message_) const
...
+        output_ = std::copy(item.username.begin(), item.username.end(), output_);
+        output_++;
+        output_ = std::copy(item.password.begin(), item.password.end(), output_);
+        output_++;
+    }
+
+    sgx_status_t sealing_status = seal_wallet(data.get(), wallet_size, (sgx_sealed_data_t*)sealed_data.get(), sealed_size);
+
+    if (sealing_status != SGX_SUCCESS){
+        on_error("[update_stored_wallet] Failed to seale wallet");
```
Comment (JoyOfMadness) [inline trusted/wallet.cpp]: If sealing failed, we should exit function here using `return;`

File: trusted/wallet.cpp [hunk 21]
```diff
@@ -215,7 +223,80 @@ void wallet::on_error(const std::string& message_) const
...
+    int ret;
+    sgx_status_t stored_status = store_wallet(&ret, sealed_data.get(), sealed_size);
+    if (ret != 0 || stored_status != SGX_SUCCESS){
+        on_error("[update_stored_wallet] Failed to to store wallet");
+    }  
 }
 
+bool wallet::load_stored_wallet()
+{
+    size_t sealed_size;
```
Comment (JoyOfMadness) [inline trusted/wallet.cpp]:
```suggestion
    size_t sealed_size{};
```

File: trusted/wallet.cpp [hunk 22]
```diff
@@ -215,7 +223,80 @@ void wallet::on_error(const std::string& message_) const
...
+        item_t item;
+        item.id = std::string{it};
+        it += item.id.size() + 1;
+        item.username = std::string{it};
+        it += item.username.size() + 1;
+        item.password = std::string{it};
+        it += item.password.size() + 1;
+        items.emplace(std::move(item));
+    }
+    _items = items;
```
Comment (JoyOfMadness) [inline trusted/wallet.cpp]:
```suggestion
    _items = std::move(items);
```

File: untrusted/enclave_wrapper.cpp [hunk 23]
```diff
@@ -187,4 +190,43 @@ void enclave_wrapper::on_error(const std::string& error_)
     std::cerr << "Error: " << error_ << std::endl;
 }
 
+int enclave_wrapper::store_wallet(const uint8_t* sealed_data, const size_t sealed_size)
+{
+    std::ofstream file(WALLET_FILE, std::ios::out | std::ios::binary);
+    if (file.fail()) {return 1;}
+    file.write((const char*) sealed_data, sealed_size);
+    file.close();
```
Comment (JoyOfMadness) [inline untrusted/enclave_wrapper.cpp]: This line is unnecessary, file gets closed on destruction

**Review by JoyOfMadness — CHANGES_REQUESTED** [review]
Some files are not formatted, try to format them using `clang-format`

File: trusted/wallet.cpp [hunk 24]
```diff
@@ -246,38 +245,38 @@ void wallet::update_stored_wallet() const
     sgx_status_t sealing_status = seal_wallet(data.get(), wallet_size, (sgx_sealed_data_t*)sealed_data.get(), sealed_size);
 
     if (sealing_status != SGX_SUCCESS){
-        on_error("[update_stored_wallet] Wallet not sealed");
+        on_error("[update_stored_wallet] Failed to seale wallet");
```
Comment (mdugacek) [inline trusted/wallet.cpp]: Sorry for typo, should be `seal` and not `sealE`.

**Review by mdugacek — APPROVED** [review]

**Review by JoyOfMadness — APPROVED** [review]
Looks good to me

**Pull request merged**